File: Octave/newton.m
function [xn, err, iter, fx] = newton(f, x0, tol, maxIter)
  % Newton-Raphson Method
	% Inputs:
	%   - f is a polinomial expression introduced as a symbolic expression
	%   - x0 is the initial value
	%   - tol is the tolerance
	% 	- maxIter is the maximum amount of iterations
	% Outputs:
	%   - xn is the solution
	%   - err is the error
	%   - iter is the amount of completed iterations
	%   - fx is f(x)
  fd = diff(f);
  f = function_handle(f);
  fd = function_handle(fd);
  xNext = x0;
  figure
  hold on
  for iter=0:maxIter
    xn = xNext;
    fdx = fd(xn);
    if fdx == 0
      disp("Error: Division by zero");
      return
    endif
    fx = f(xn);
    xNext = xn - fx/fdx;
    err = abs(xNext - xn) / abs(xNext);
    if err <= tol
      break
    endif
    plot(iter, err, 'ro');
    title("Newton-Raphson Method");
    xlabel("Iterations")
    ylabel("Error")
  endfor
endfunction

File: Python/newton.py
from sympy import *

x = symbols('x')


def newton(expr, x0, tol, maxIter):
    """ Newton Method

    Arguments:
        expr {string} -- is the polinomial expression
        x0 {float} -- is the initial value x_0
        tol {float} -- is the tolerance
        maxIter {int} -- is the max amount of iterations
    Returns:
        xn {float} -- is the solution
        err {float} -- is the error
        _iter {int} -- is the amount of iterations
        fx {float} -- is f(xn)

    """
    errReturn = [0, 0, 0, 0]
    try:
        f = sympify(expr)
        fd = diff(expr, x)
        xNext = sympify(x0)
        for _iter in range(0, maxIter):
            xn = N(xNext)
            fdx = fd.subs(x, xn)
            if (fdx == 0):
                print("Error: Division by zero")
                return errReturn
            fx = f.subs(x, xn)
            xNext = xn - fx/fdx
            err = abs(xNext - xn)/abs(xNext)
            if (err <= tol):
                break
        return [N(xn), N(err), _iter, N(fx)]
    except:
        print("There was an error.")
        return errReturn
